File: app/src/main/java/com/azizbekian/spyur/BaseApplication.java
package com.azizbekian.spyur;

import android.content.Context;
import android.support.multidex.MultiDexApplication;

import com.azizbekian.spyur.injection.AppComponent;
import com.azizbekian.spyur.injection.AppModule;
import com.azizbekian.spyur.injection.DaggerAppComponent;

/**
 * Created on July 29, 2016.
 *
 * @author Andranik Azizbekian (andranik.azizbekyan@gmail.com)
 */
public class BaseApplication extends MultiDexApplication {

    private static AppComponent sAppComponent;

    @Override
    public void onCreate() {
        super.onCreate();

        sAppComponent = DaggerAppComponent.builder()
                .appModule(new AppModule(this))
                .build();
    }

    public static AppComponent getComponent() {
        return sAppComponent;
    }

    public static Context getContext() {
        return sAppComponent.getApplicationContext();
    }
}

File: app/src/main/kotlin/com/incentive/yellowpages/ui/base/BaseApplication.kt
package com.incentive.yellowpages.ui.base

import android.content.Context
import android.support.multidex.MultiDexApplication
import com.bumptech.glide.request.target.ViewTarget
import com.incentive.yellowpages.BuildConfig
import com.incentive.yellowpages.R
import com.incentive.yellowpages.injection.component.AppComponent
import com.incentive.yellowpages.injection.component.DaggerAppComponent
import com.incentive.yellowpages.injection.module.AppModule
import timber.log.Timber

open class BaseApplication : MultiDexApplication() {

    companion object {
        @JvmStatic lateinit var appComponent: AppComponent
        lateinit var context: Context
    }

    override fun onCreate() {
        super.onCreate()

        // workaround for Glide issue - https://goo.gl/sLdXQx
        ViewTarget.setTagId(R.id.glide_tag)

        appComponent = DaggerAppComponent.builder()
                .appModule(AppModule(this))
                .build()

        context = this
        plantTimber()
    }

    private fun plantTimber() {
        if (BuildConfig.DEBUG) {
            Timber.plant(Timber.DebugTree())
        }
    }

}
